src/mars.c

```diff
@@ -1694,21 +1710,74 @@ Language changes listed by -transition=id:\n\
                 fprintf(global.stdmsg, "code      %s\n", m->toChars());
 
             obj_start(m->srcfile->toChars());
-            genObjFile(m, global.params.multiobj);
+            genObjFile(m, false);
             if (entrypoint && m == rootHasMain)
-                genObjFile(entrypoint, global.params.multiobj);
+                genObjFile(entrypoint, false);
             for (size_t j = 0; j < Module::amodules.dim; j++)
             {
+                // todo: This part is not the best for the total size of object files.
+                // For example:
+                // When both of main.d and code.d import mx, even if main.obj
+                // already contains mx->marray, it is stored in code.obj again.
                 Module *mx = Module::amodules[j];
-                if (mx != m && mx->importedFrom == m && (mx->marray || mx->massert || mx->munittest))
-                    genhelpers(mx, true);
```

Comment:
Generating `__unittest_fail` is just a bug. I'm not adding improvement for that in this PR, but I can easily kill it.